### DISCUSSION TITLE
Discussion: Namespace Access Modifier

### DISCUSSION THREAD

**Discussion by brainwipe** [dotnet/csharplang#3870] [General] [answered]
**Purpose:** 
Allow class access to be restricted to the current namespace or it's children.

**Example:**
I introduce the new access modifier "spaced" (can't think of a better name (and I like the 90s show)) because I want to show developers that they shouldn't be using BaseSurvey outside of that namespace.

```csharp
namespace Company.Surveys
{
    internal interface ISurvey {}

    spaced class BaseSurvey : ISurvey {}

    internal class HealthCheckSurvey : ISurvey {}
}

namespace Company.Patient
{
    internal class Patient
    {
         public ISurvey CreateSurvey()
         {
              return new HealthCheckSurvey(); // Good
              return new BaseSurvey(); // BAD!
         }
    }
}

```

**Benefit**
- Logical separation of domain areas in a single assembly

That's probably it!

**Notes**
I am trying to partition a monolith code base into neat domain areas and we are finding that there are groups of classes that belong to each other but not outside of that domain. We want to avoid splitting the solution into many assemblies because if we need to merge them again then that's a lot of effort. For some specific classes, we want to say "don't use this outside this namespace".

Many thanks for your time and consideration!

**Comment by ulln** [comment, marked as answer]
Related: [namespace internal access modifier](https://github.com/dotnet/csharplang/issues/2497)

**Comment by jnm2** [comment]
You could do this independently of the language with a custom `[Spaced]` attribute and an analyzer to go with it.
